pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/ConsumptionDelayTracker.java

```diff
@@ -0,0 +1,295 @@
+  // ideal state. This is done with the goal of minimizing reading ideal state for efficiency reasons.
+  // _partitionsMarkedForVerification<Key=PartitionGroupId,Value=TimePartitionWasMarkedForVerificationMilliseconds>
+  private ConcurrentHashMap<Long, Long> _partitionsMarkedForVerification = new ConcurrentHashMap<>();
+  // Sleep interval for timer thread that triggers read of ideal state
+  private static final int TIMER_THREAD_TICK_INTERVAL_MS = 300000; // 5 minutes +/- precision in timeouts
+  // Once a partition is marked for verification, we wait 10 minutes to pull its ideal state.
+  private static final int PARTITION_TIMEOUT_MS = 600000;          // 10 minutes timeouts
+  // Delay Timer thread for this time after starting timer
+  private static final int INITIAL_TIMER_THREAD_DELAY_MS = 1000;
+  // Mutable versions of the above constants so we can test with smaller delays
+  final int _timerThreadTickIntervalMs;
+  final int _initialTimeThreadDelayMs;
+  // Timer task to check partitions that are inactive against ideal state.
+  final private Timer _timer;
+
+  final private ServerMetrics _serverMetrics;
+  final private String _tableNameWithType;
+
+  private boolean _enableAging;
+  final private Logger _logger;
+
+  final private RealtimeTableDataManager _realTimeTableDataManager;
+
+  /*
+   * Helper function to update the maximum when the current maximum is removed or updated.
+   * If no samples left we set maximum to minimum so new samples can be recorded.
+   */
+  private DelayMeasure getMaximumDelay() {
+    DelayMeasure newMax = null;
+    for (long pid : _partitionToDelaySampleMap.keySet()) {
+      DelayMeasure currentMeasure = _partitionToDelaySampleMap.get(pid);
+      if ((newMax == null) || (currentMeasure._delayMilliseconds > newMax._delayMilliseconds)) {
+        newMax = currentMeasure;
+      }
+    }
+    return newMax;
+  }
+
+  private List<Integer> getPartitionsHostedByThisServerPerIdealState() {
+    return _realTimeTableDataManager.getTablePartitionsHostedByThisServer();
+  }
+  /*
+   * Helper function to be called when we should stop tracking a given partition. Removes the partition from
+   * all our maps, it also updates the maximum if the tracked partition was the previous maximum.
+   *
+   * @param partitionGroupId partition ID which we should stop tracking.
+   */
+  private void removePartitionId(long partitionGroupId) {
+    _partitionToDelaySampleMap.remove(partitionGroupId);
+    // If we are removing a partition we should stop reading its ideal state.
+    _partitionsMarkedForVerification.remove(partitionGroupId);
+   }
+
+  // Custom Constructor
+  public ConsumptionDelayTracker(ServerMetrics serverMetrics, String tableNameWithType,
+      RealtimeTableDataManager realtimeTableDataManager, int initialTimeThreadDelayMs, int timerThreadTickIntervalMs)
+      throws RuntimeException {
+    _logger = LoggerFactory.getLogger(tableNameWithType + "-" + getClass().getSimpleName());
```

Comment:
raw table name should suffice.

Reply:
I think these names are more telling and will help in debugging, so leaving as is. Same for the one below.

Reply:
Why do you think having the "_REALTIME" will help here?

Reply:
you never know when we may intro a new timer for the other part right?

Reply:
Log lines show up being really long with threadname. This class will not be associated with anything in the offline table. If at a later day, we do see the need for it, we can change the logger name.
please use raw table name